sound/soc/sof/pm.c

```diff
@@ -411,6 +411,28 @@ int snd_sof_runtime_resume(struct device *dev)
 }
 EXPORT_SYMBOL(snd_sof_runtime_resume);
 
+int snd_sof_set_d0_substate(struct snd_sof_dev *sdev,
+			    enum sof_d0_substate d0_substate)
+{
+	int ret;
+
+	/* Todo: destroy/reload unused pipelines/widgets in FW */
+
+	/* do platform specific set_state */
+	ret = snd_sof_dsp_set_power_state(sdev, d0_substate);
+	if (ret < 0) {
+		dev_err(sdev->dev,
+			"error: set_power_state error %d\n", ret);
+		return ret;
+	}
+
+	/* update dsp D0 sub-state */
+	sdev->d0_substate = d0_substate;
+
+	return 0;
+}
+EXPORT_SYMBOL(snd_sof_set_d0_substate);
```

Comment:
@keyonjie the commit message says "it will destroy unused pipelines" but it is called out as a TODO. If I were you I'd remove both the commit message that says destroy unused pipelines and the TODO. We can always enhance this later.